mdp/nodes/sfa_nodes.py

```diff
@@ -134,22 +169,25 @@ def _train(self, x, include_last_sample=None):
 
     def _stop_training(self, debug=False):
         ##### request the covariance matrices and clean up
-        self.cov_mtx, self.avg, self.tlen = self._cov_mtx.fix()
-        del self._cov_mtx
+        if hasattr(self, '_dcov_mtx'):
+            self.cov_mtx, self.avg, self.tlen = self._cov_mtx.fix()
+            del self._cov_mtx
         # do not center around the mean:
         # we want the second moment matrix (centered about 0) and
         # not the second central moment matrix (centered about the mean), i.e.
         # the covariance matrix
-        self.dcov_mtx, self.davg, self.dtlen = self._dcov_mtx.fix(center=False)
-        del self._dcov_mtx
+        if hasattr(self, '_dcov_mtx'):
+            self.dcov_mtx, self.davg, self.dtlen = self._dcov_mtx.fix(center=False)
+            del self._dcov_mtx
 
         rng = self._set_range()
 
         #### solve the generalized eigenvalue problem
         # the eigenvalues are already ordered in ascending order
         try:
-            self.d, self.sf = self._symeig(self.dcov_mtx, self.cov_mtx,
-                                     range=rng, overwrite=(not debug))
+            self.d, self.sf = self._sfa_solver(
+                    self.dcov_mtx, self.cov_mtx, True, "on", rng,
+                    overwrite=(not debug))
```

Comment:
In my current revision this part looks as follows:
```
        # solve the generalized eigenvalue problem
        if self._eig_solve is 'symeig':
            self.d, self.sf = self._eig_solve_symeig( rng, debug )
        elif self._eig_solve is 'svd':
            self.d, self.sf = self._eig_solve_svd( rng )
```
I find this a bit cleaner than "hiding" the calls behind the non-descript `_sfa_solver()` call. To me it's okay if the different methods have different signatures, as they are, well, different methods. Forcing the same signature on all of them feels counter productive to me. While I can see how this approach enable the easy switching of methods, I'd still prefer a more explicit style that's easier to follow for newcomers facing the code.

Reply:
As I perceived it, it is kind of an MDP requirement that signatures are fully compatible. I changed it to call-by-order rather than call-by-name, because some of the argument names hide global names, i.e. `type` and `range`. A former version of one of my solvers used `range` for iteration, so I wanted to be able to rename the range parameter, which wouldn't be possible in call-by-name case. Now it's not necessary any more, still it is bad style to hide global names and the call should allow implementations to choose different names. That's why call-by-order might be better here.

Reply:
Signatures for the same procedures, probably yes. My point would be that the different methods on offer are not the same procedures and thus shouldn't be treated as such.

Reply:
In https://github.com/mdp-toolkit/mdp-toolkit/pull/30#discussion_r136691665 pberkes stated that
> Nothing else in _stop_training needs to change.

I took that as a hint to avoid a case distinction inside of `stop_training`.

Reply:
I think the point @pberkes was making is that it is easily possible to put every method behind the same signature and set the chosen approach in the `__init__`, leaving `_stop_training` completely untouched.

While this is a very efficient way of implementing it, my personal preference would be an explicit selection from the available eigenvalue solvers. (Though when we're at the point of arguing personal preferences, then we might as well go with any working solution; that's usually not time well spent.)